api/store.js

```diff
@@ -0,0 +1,175 @@
+          logger.silly(`blob [${key}] does ${exists ? '' : 'not '}exist`);
+          resolve(exists);
+        }
+      });
+    });
+  }
+
+  /**
+   * @async
+   * Gets the data from a specified blob.
+   * @param {String} key - Blob identifier
+   * @returns {Buffer} A buffer of the data in the blob.
+   */
+  async get(key) {
+    return new Promise((resolve, reject) => {
+      logger.silly(`getting data from blob [${key}]`);
+      // TODO - make this a buffer or byte array, but not a string
+      let out = '';
+      const readStream = this.store.createReadStream(key);
+      readStream.on('data', chunk => {
+        out += chunk;
+      });
+      readStream.on('end', () => {
+        resolve(out);
+      });
+      readStream.on('error', e => {
+        logger.error(e);
+        reject(e);
+      });
+    });
+  }
+
+  /**
+   * Get a readable stream for a specified blob.
+   * @param {String} key - Blob identifier
+   * @returns {ReadableStream} A readable stream
+   */
+  getReadStream(key) {
+    logger.silly(`getting a read stream for blob [${key}]`);
+    return this.store.createReadStream(key);
+  }
+
+  /**
+   * @async
+   * Remove a blob
+   * @param {String} key - Blob identifier
+   */
+  async remove(key) {
+    return new Promise(resolve => {
+      logger.silly(`removing blob [${key}]`);
+      this.store.remove(key, () => resolve());
+    });
+  }
+
+  /**
+   * @async
+   * Write data into a blob with a specified key.  The blob
+   * will be created if it does not already exist.
+   * @param {String} key - Blob identifier
+   * @param {(ReadableStream|*)} - The data to write.  If a ReadableStream is
+   *    provided, its contents will be piped into the blob.  For everything
+   *    else, it will be written directly to the blob.
+   * @param {*} opts - Returned value when the write is finished.
+   */
+  async write(key, data, opts) {
+    return new Promise((resolve, reject) => {
+      logger.silly(`writing to blob [${key}]`);
+      const writeStream = this.store.createWriteStream(key);
+      let rejected = false;
+
+      const done = () => {
+        logger.silly('finished writing blob');
+        if (!rejected) {
+          resolve(opts);
+        }
+      };
+
+      const errHandler = e => {
+        rejected = true;
+        logger.error(e);
+        reject(e);
+      };
+
+      if (isReadableStream(data)) {
+        logger.silly('source is a stream');
+        data.pipe(writeStream);
+
+        data.on('close', () => {
+          writeStream.end();
+          done();
+        });
+
+        data.on('error', errHandler);
+        writeStream.on('error', errHandler);
+      } else {
+        logger.silly('source is memory data');
+        writeStream.write(data);
+        writeStream.end(done);
+        writeStream.on('error', errHandler);
+      }
+    });
+  }
+}
+
+const def = new Store(
+  getBlobStore(process.env.STORE_TYPE || '', process.env.STORE_PATH || '')
+);
+
+Store.exists = def.exists.bind(def);
```

Comment:
i'm not sure i follow this class method binding -- can you walk me through what it accomplishes?

Reply:
can we not just export both the `def` instance as well as the `Store` class?

Reply:
The binding is a result of how the methods are being exported.  Without the binding, when they're called, their `this` is set to the `Store` class itself, not the instance, so `this.store` resolves `undefined` and everything breaks.

I'm not fundamentally opposed to two exports.  Do you mean something like this?

```javascript
module.exports = {
  default: def,
  class: Store
};
```

My thinking was just that it's nice and clean from a module consumer's perspective to be able to just import a thing and use it.  Ultimately the whole reason for having a class at all is so I can inject the store for testing, and then this import stuff is just to make the module easy to consume in actual use.  Ideally we'd never use the constructor except for the default and for testing.

I'm totally open to other approaches for this, though!